module/os/linux/zfs/zfs_ctldir.c

```diff
@@ -1102,6 +1137,53 @@ zfsctl_snapshot_mount(struct path *path, int flags)
 		goto error;
 	}
 
+#if defined(USE_FC_MOUNT)
+	(void) &zfsctl_snapshot_alloc; // For now..
+	error = -zfsctl_snapshot_path(path, MAXPATHLEN, full_path);
+	if (error)
+		goto error;
+
+	struct fs_context *fc;
+	struct vfsmount *mnt = NULL;
+	fc = fs_context_for_submount(&zpl_fs_type, dentry);
+	if (IS_ERR(fc)) {
+		zfs_dbgmsg("error-fc: %p", ERR_CAST(fc));
+	} else {
+		vfs_parse_fs_string(fc, "source", full_name, strlen(full_name));
+		mnt = fc_mount(fc);
+
+		if (IS_ERR(mnt)) {
+			zfs_dbgmsg("error-mnt: %p", ERR_CAST(mnt));
+		}
+		zfs_dbgmsg("mnt %p", mnt);
+		zfs_dbgmsg("put_fc %p", fc);
+		put_fs_context(fc);
+	}
+
+	if (!mnt || IS_ERR(mnt)) {
+		error = PTR_ERR(mnt);
+		goto error;
+	} else {
+		zfs_dbgmsg("yielding vfsmount %p", mnt);
+		mnt->mnt_flags |= MNT_SHRINKABLE;
+		mntget(mnt);
+		printk("adding mount to expiry list with timeout %d.",
+		    zfs_expire_snapshot);
+		spl_add_mount_to_expire(mnt, zfs_expire_snapshot);
```

Comment:
zfs_expire_snapshot <= 0 means 'do not expire' but here passing 0 to spl_add_mount_to_expire() seems to expire the mount anyway. It is very important to never expire a snapshot mount when used over nfs.

Reply:
Interesting, I was not aware of this semantics of `zfs_expire_snapshot`, I think we should add a test case and preserve that semantics.

However I don't understand how this is related to usage over nfs. Shouldn't the snapshots just get automounted/autounmounted over nfs like they do regularly?

Reply:
Once the snapshot is accessed over nfs, if it is auto-unmounted, the nfs client will get 'stale file handle' and other strange error when trying to access the snapshot again.

Reply:
So currently if you want to expose auto-mounted snapshots via nfs you always configure `zfs_expire_snapshot = 0`? Maybe that behavior needs separate investigation, as I would consider that to be a bug as well.

Reply:
Correct. That's how we ended up doing. I wouldn't say it is a bug, what can you expect when a cross mount (snapshot) is gone (unmounted) from the server?

Reply:
I would expect it being re-mounted when it is being accessed. Actually the nfs mount (i.e. the nfs client) should also expire after not being access for some time. I've experimented with the nfs mount and observed the following behavior:
State: ds mounted, snap not mounted, nfs mounted, nfs-snap not moutned
Action: ls on nfs-snap, success
State: ds mounted, snap mounted, nfs mounted, nfs-snap mounted
Action: umount snap, fails with "busy"
State: ds mounted, snap mounted, nfs mounted, nfs-snap mounted
Action: Wait for more than 10mins
State: ds mounted, snap not mounted, nfs mounted, nfs-snap not mounted
Action: ls on nfs-snap, failure (stale file handle)
I would have expected: success and state: ds mounted, snap mounted, nfs mounted, nfs-snap mounted.

I would (naively) expect that you cannot unmount the snap while a client still has the snapshot via nfs automounted and once all clients have expired the server can expire as well.